### ISSUE TITLE
Element being a list causes tests to randomly break

### ISSUE THREAD

**Issue by Lugribossk** [bruderstein/unexpected-react#6]
This test below (a simplified version of a real test in my codebase) causes an error when run. However the error seems to happen after the test has run, so the test will still succeed. Randomly (depending on timing I guess) it might break a later test in the test suite which will then show the stack trace:
```js
TypeError: Cannot read property 'displayName' of undefined
    at ReactElementAdapter.getName (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike-jsx-adapter\lib\ReactElementAdapter.js:72:32)
    at convertToDiff (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\lib\convertToDiff.js:29:23)
    at C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\lib\convertToDiff.js:46:24
    at Array.map (native)
    at convertToDiff (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\lib\convertToDiff.js:45:43)
    at C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\lib\diff.js:487:69
    at Array.forEach (native)
    at C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\lib\diff.js:468:21
    at C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:224:17
    at module.exports.setEqual (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:199:28)
    at module.exports.setEqual (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:220:24)
    at module.exports.setEqual (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:220:24)
    at module.exports.setEqual (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:220:24)
    at module.exports.setEqual (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:220:24)
    at C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:223:13
    at [object Object]._onTimeout (C:\Code\Sandbox\node_modules\unexpected-react\node_modules\unexpected-htmllike\node_modules\array-changes-async\lib\arrayChanges.js:134:21)
```

```js
import React from "react";
import ReactTestUtils from "react-addons-test-utils";
import expect from "unexpected"; // unexpected-react plugin has been installed somewhere else

class Example extends React.Component {
    renderDetails(item) {
        return (
            <div>
                <h1>Test</h1>
                {item.details.map(label => <h2 key={label}>{label}</h2>)}
                <h3>Test4</h3>
            </div>
        );
    }

    render() {
        return (
            <div>
                {this.props.items.map(item => this.renderDetails(item))}
            </div>
        );
    }
}

describe("Example", () => {
    it("should render item details.", () => {
        let renderer = ReactTestUtils.createRenderer();
        renderer.render(
            <Example items={[{details: ["Test2", "Test3"]}]} />
        );

        expect(renderer, "to contain", (
            <div>
                <h1>Test1</h1>
                <h2>Test2</h2>
                <h2>Test3</h2>
                <h3>Test4</h3>
            </div>
        ));
    });
});
```

Due to the random nature of it I haven't been able to reproduce it failing with an error consistently, but the problem can be seen by editing `ReactElementAdapter` in `unexpected-htmllike-jsx-adapter` like this. It will show that `element` is a list in this case. A similar problem happens in getChildren() where the element does not have a props property.
```js
class ReactElementAdapter {
    //...

    getName(element) {
        if (typeof element.type === 'string') {
            return element.type;
        }
        if (!element.type) {
            console.log(element)
        }

        return element.type.displayName || element.type.name || 'no-display-name';
    }

    //...
}
```

Perhaps it would also be good to ensure that things do not happen asynchronously after tests, i.e. make everything unexpected-react does synchronous? That would also make it easier to mix testing of normal rendering and shallow rendering, without needing to remember to return the expect call when using the normal renderer. That confused me a bit when migrating from `unexpected-react-shallow`.

Oh and thanks a lot for writing this plugin, it really makes testing React components a breeze!

**Comment by bruderstein** [comment]
Many thanks for the report.

PR #5 shows this too.  I'm writing a fix!

I'm also working on a reasonably major rework to make the assertions synchronous when they can be, and fall back to async when needed. 

They are only async when you add asynchronous `expect.it()` assertions into your JSX (this is a new feature over unexpected-react-shallow)

e.g. a component that gets passed a promise in a prop
```js
return expect(component, 'to have rendered', <SomeComponent data={ expect.it('to be resolved with', 'some value') } />);
```

This means you can do cool things like using `unexpected-resemble` to compare a generated data-url image with a reference image etc etc.

I'll get the fix for the lists in the next few days, and probably the sync/async changes in the new year.

**Comment by Lugribossk** [comment]
Thanks, that sounds great!

That example with the embedded assertion does look quite cool. The thing that worries me with the async assertions is that it's easy to forget to return them (especially if they're mixed in with test cases that don't assert on the rendered output) and when you do the test won't fail because of it.

It would be nice if there was some way to warn the user that they're just throwing an async assertion away but I guess there isn't really a way to do that which is test runner agnostic.

**Comment by bruderstein** [comment]
For mocha, unexpected throws an error if you forget to return the result of the expect. If you're not seeing that, or are using another runer, it'd be really helpful if you could let us know what you're using, how you're running it (via npm, gulp/grunt, command line etc), and which versions.

**Comment by Lugribossk** [comment]
I'm using Mocha, and not seeing errors, so I'll try to reproduce the essential parts of my setup and report a bug for it. Should I do that on the unexpected project itself?

**Comment by bruderstein** [comment]
Yes, that should be reported to the unexpected project - tell 'em I sent ya! :)

The actual issue is now fixed in 0.4.0 - thanks again for reporting.
